src/dotty/tools/dotc/core/Skolemization.scala

```diff
@@ -0,0 +1,127 @@
+package dotty.tools.dotc
+package core
+
+import Symbols._, Types._, Contexts._
+import collection.mutable
+
+/** Methods to add and remove skolemtypes.
+ *  
+ *  Skolem types are generated when comparing refinements. 
+ *  Two skolem types are equal if they refer to the same underlying type. 
+ *  To avoid unsoundness, skolem types have to be kept strictly local to the
+ *  comparison, they are not allowed to escape the lifetime of a comparison
+ *  by surviving in a context or in GADT bounds. 
+ */
+trait Skolemization {
+  
+  implicit val ctx: Context
+
+  protected var skolemsOutstanding = false
+  
+  def ensureSingleton(tp: Type): SingletonType = tp.stripTypeVar match {
+    case tp: SingletonType => 
+      tp
+    case tp: ValueType => 
+      skolemsOutstanding = true
+      SkolemType(tp)
+    case tp: TypeProxy => 
+      ensureSingleton(tp.underlying)
+  }
+    
+  /** Approximate a type `tp` with a type that does not contain skolem types.
+   *  @param  toSuper   if true, return the smallest supertype of `tp` with this property
+   *                    else return the largest subtype.
+   */
+  final def deSkolemize(tp: Type, toSuper: Boolean): Type = 
+    if (skolemsOutstanding) {
+      skolemsOutstanding = false
```

Comment:
Could we avoid using side-effects here? Or rename `deSkolemize` to something that makes it very clear that you can't just use it whenever you want to deskolemize something without breaking the rest of the program.